### PR TITLE
Add Region and Signer Algorithm Overrides to S3 Repos

### LINKED ISSUE #51861: Add the Ability to Specify Region and Signer Override in S3 SDK
Some S3 compatible blob stores force v4 authentication with regions other than the default `us-east-1` or don't support v4 authentication.
Currently there is no way of changing the signing region (we only allow setting the endpoint) or forcing the S3 SDK to not use v4 auth if the endpoint does not support it.

We should bring back the region setting to allow v4 auth to properly work if the blobstore supports it.
Also, we should expose the [signer override setting](https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/ClientConfiguration.html#setSignerOverride-java.lang.String-) to allow forcing the supported auth mechanism in case the SDK is unable to determine the correct one for a given endpoint. 

Link: https://discuss.elastic.co/t/s3-snapshot-on-s3-compatible-servers/217750/4

### PR DESCRIPTION
**Pull request by original-brownbear**
Exposes S3 SDK signing region and algorithm override settings as requested in #51861.

Closes #51861

### CODE REVIEW

File: docs/plugins/repository-s3.asciidoc [hunk 1]
```diff
@@ -184,6 +184,22 @@ pattern then you should set this setting to `true` when upgrading.
     https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/s3/AmazonS3Builder.html#disableChunkedEncoding--[AWS
     Java SDK documentation] for details. Defaults to `false`.
 
+`region`::
+
+    Allows specifying the signing region to use. Specificing this setting manually should not be necessary for most use cases. Generally,
+    the SDK will correctly guess the singing region to use. It should be considered an expert level setting to support S3-compatible APIs
```
Comment (tlrx) [inline docs/plugins/repository-s3.asciidoc]: Typo `singing`

File: docs/plugins/repository-s3.asciidoc [hunk 2]
```diff
@@ -184,6 +184,22 @@ pattern then you should set this setting to `true` when upgrading.
     https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/s3/AmazonS3Builder.html#disableChunkedEncoding--[AWS
     Java SDK documentation] for details. Defaults to `false`.
 
+`region`::
```
Comment (tlrx) [inline docs/plugins/repository-s3.asciidoc]: I'm wondering if it should be named `signing_region` to make it more explicit?

**Review by tlrx — COMMENTED** [review]
I've left two minor comments. I'm a bit more concerned about how we test these settings, can we have an integration test for this?

File: docs/plugins/repository-s3.asciidoc (hunk 2 above)
Comment (ywelsch) [reply docs/plugins/repository-s3.asciidoc]: I prefer to use the same naming convention as AWS does (i.e. just plain "region")

**Comment by original-brownbear** [comment]
Thanks for taking a look Tanguy!

> an we have an integration test for this?

Sure we can, it's just a matter of coding it up :) I wonder if it's worth the effort? It seems that this would be outright testing the SDK.
I realise that we do this for retrying, but there we are party to things also in providing a resettable input stream. Here, we'd really just be testing the SDK since we do verify that the client is build with the correct settings in the current tests IMO.

The amount of code it takes to add these tests makes me not want to do it. The fact that SDKs have bugs does makes me want to add the test :D 
Tbh. I don't think it's worth it adding it now but maybe we could add a TODO/issue asking to enhance our tests to cover the authorisation side of things as well for S3 and deal with it at lower priority?

**Comment by DaveCTurner** [comment]
I dunno, I think there's value in an integration test too, just checking that we've got all the end-to-end plumbing right. Maybe against Minio with an explicit region, given that the point of this feature is to support non-S3 services with nonempty regions?

**Comment by original-brownbear** [comment]
> Maybe against Minio with an explicit region, given that the point of this feature is to support non-S3 services with nonempty regions?

The annoying thing with that is that it forces us to add yet another Minio IT run and I'm not sure you can really force Minio to only accept v4 signatures. So we can only test signer override + region.

But I just realised we have no auth testing whatsoever on our S3 mock ... probably good to have it in some form, let's add it for this effort then :) On it.

File: docs/plugins/repository-s3.asciidoc (hunk 2 above)
Comment (tlrx) [reply docs/plugins/repository-s3.asciidoc]: Ok

**Comment by tlrx** [comment]
> Sure we can, it's just a matter of coding it up :) I wonder if it's worth the effort? It seems that this would be outright testing the SDK.

Yes I know :( But we have to maintain such settings for a long time and we previously broke them multiple times without noticing, mostly because we were lacking integration tests. This would also be easier to migrate to AWS SDK V2 if we have more tests. I completely understand your point but I think it will still be valuable at the end.

**Comment by DaveCTurner** [comment]
> But I just realised we have no auth testing whatsoever on our S3 mock 

Good catch. We used to, but it looks like we lost them [here](https://github.com/elastic/elasticsearch/pull/49107/files#diff-6f84ca24bf0f20c7c9d7a3e0d9792f10L148-L169) in #49107.

**Comment by DaveCTurner** [comment]
> Good catch. We used to, but it looks like we lost them here in #49107.

Wait, I found them again. This logic got split up across `S3HttpFixture` and its subclasses. Sorry for the noise. Maybe you mean some stronger auth testing?

**Comment by original-brownbear** [comment]
> Wait, I found them again. This logic got split up across S3HttpFixture and its subclasses. Sorry for the noise. Maybe you mean some stronger auth testing?

No worries, I was also confused by this and was only referring to the integration tests.

For now I decided to keep it simple though and added the minimum testing required to verify that both setting enforce the right region and/or signer override in https://github.com/elastic/elasticsearch/pull/52112/commits/d54c45fecf0454a39650a6a395f1d407c6771e8d => let me know what you think :)

**Comment by rpasche** [comment]
I can confirm that this patch is working. I was able to backport it to 7.4.1 and able to test this plugin on one of our testclusters against an openstack swift container. Was able to run backup successfully. Would really love to see this included in v7.7. Thanks.

**Review by ywelsch — APPROVED** [review]
LGTM

**Comment by original-brownbear** [comment]
Thanks Yannick + Tanguy!

**Pull request merged**